File: ql/time/calendars/target.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
*/

#include <ql/time/calendars/target.hpp>

namespace QuantLib {

    TARGET::TARGET() {
        // all calendar instances share the same implementation instance
        static std::shared_ptr<Calendar::Impl> impl(new TARGET::Impl);
        impl_ = impl;
    }

    bool TARGET::Impl::isBusinessDay(const Date& date) const {
        Weekday w = date.weekday();
        Day d = date.dayOfMonth(), dd = date.dayOfYear();
        Month m = date.month();
        Year y = date.year();
        Day em = easterMonday(y);
        if (isWeekend(w)
            // New Year's Day
            || (d == 1  && m == January)
            // Good Friday
            || (dd == em-3 && y >= 2000)
            // Easter Monday
            || (dd == em && y >= 2000)
            // Labour Day
            || (d == 1  && m == May && y >= 2000)
            // Christmas
            || (d == 25 && m == December)
            // Day of Goodwill
            || (d == 26 && m == December && y >= 2000)
            // December 31st, 1998, 1999, and 2001 only
            || (d == 31 && m == December &&
                (y == 1998 || y == 1999 || y == 2001)))
            return false; // NOLINT(readability-simplify-boolean-expr)
        return true;
    }

}

File: quantlib/time/calendars/target.py
from quantlib.time.calendar import Calendar
from quantlib.time.date import Date
from quantlib.time.month import Month


class TARGET(Calendar):
    """TARGET calendar (Trans-European Automated Real-time Gross Settlement Express Transfer)."""

    class _Impl(Calendar.WesternImpl):
        def name(self) -> str:
            return "TARGET"

        def isBusinessDay(self, d: Date) -> bool:
            w = d.weekday()
            day = d.dayOfMonth()
            dd = d.dayOfYear()
            m = d.month()
            y = d.year()
            em = self.easterMonday(y)

            if (
                self.isWeekend(w)
                or (day == 1 and m == Month.January)
                or (dd == em - 3 and y >= 2000)  # Good Friday
                or (dd == em and y >= 2000)  # Easter Monday
                or (day == 1 and m == Month.May and y >= 2000)  # Labour Day
                or (day == 25 and m == Month.December)  # Christmas
                or (day == 26 and m == Month.December and y >= 2000)  # Day of Goodwill
                or (day == 31 and m == Month.December and y in (1998, 1999, 2001))
            ):
                return False
            return True

    def __init__(self):
        super().__init__(TARGET._Impl())
